### PR TITLE
add: draft SSH Keypair management feature in usersettings(#657)

### PR DESCRIPTION
**Pull request by lizable**
This is on-going issue about applying SSH Keypair UI.
UI flow is almost finished, For now.
But unfortunately, There are some vague understanding about receiving ssh public-key and sending a request to make a random ssh keypair (both public and private key).
It would be appreciated anyone kindly give me some reviews or any comments about this.

### CODE REVIEW

File: src/components/backend-ai-usersettings-general-list.ts [hunk 1]
```diff
@@ -610,6 +671,13 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
     this._hideCurrentEditorChangeDialog();
   }
 
+  _copySSHKey(keyName : string) {
+    if (keyName !== "") {
+      let copyText = this.shadowRoot.querySelector(keyName);
+      navigator.clipboard.writeText(copyText.textContent);
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: This line raises an error in my browser. Please check if this universally works.
image

File: index.html [hunk 2]
```diff
@@ -25,7 +25,7 @@
 
     globalThis.process = {env: {NODE_ENV: 'production'}};
     globalThis.packageVersion = "20.08.0";
-    globalThis.buildVersion = "200804.090818";
+    globalThis.buildVersion = "200806.110838";
```
Comment (adrysn) [inline index.html]: I think `buildVersion` should be changed when a new build/release is published. Please revert this change.

File: src/components/backend-ai-console.ts [hunk 3]
```diff
@@ -999,7 +999,7 @@ export default class BackendAIConsole extends connect(store)(LitElement) {
             </div>
             <address>
               <small class="sidebar-footer">Lablup Inc.</small>
-              <small class="sidebar-footer" style="font-size:9px;">20.08.0.200804</small>
+              <small class="sidebar-footer" style="font-size:9px;">20.08.0.200806</small>
```
Comment (adrysn) [inline src/components/backend-ai-console.ts]: Same here.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 4]
```diff
@@ -800,6 +877,57 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
                      ${_t("button.Save")}</wl-button>
         </div>
       </backend-ai-dialog>
+      <backend-ai-dialog id="ssh-keypair-management-dialog" fixed backdrop persistent>
+      <span slot="title">${_t("usersettings.SSHKeypairManagement")}</span>
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: Please indent one-level here.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 5]
```diff
@@ -800,6 +877,57 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+              <wl-icon>content_copy</wl-icon>
+            </wl-button>
+          </div>
+          <span slot="title">${_t("usersettings.PrivateKey")}</span>
+          <div class="horizontal">
+              <pre id="ssh-private-key" style="max-width:50ch; white-space:nowrap; overflow:scroll; display:inline-block;"></pre>
+              <wl-button class="copy" @click="${() => this._copySSHKey("#ssh-private-key")}" style="display:inline-block;">
+              <wl-icon>content_copy</wl-icon>
+            </wl-button>
+          </div>
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: Please amend indents here.

File: version.json [hunk 6]
```diff
@@ -1 +1 @@
-{ "package": "20.08.0", "build": "200804.090818", "revision": "4050ebf0" }
+{ "package": "20.08.0", "build": "200806.110838", "revision": "0dbded10" }
```
Comment (adrysn) [inline version.json]: This also better not be changed.

**Comment by inureyes** [comment]
The changes ignore too many listing in .editorconfig.  (e.g. json should use 2 spaces for indent)

**Comment by inureyes** [comment]
Fixed indentation problems. @lizable  please fetch / pull the branch :)

File: src/components/backend-ai-usersettings-general-list.ts [hunk 7]
```diff
@@ -81,6 +83,13 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
           margin-right: 5px;
         }
 
+        span[slot="title"] {
+          font-weight: bold;
+          margin-top: 15px !important;
+          margin-bottom: 15px;
+          display:inline-block;
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: Minor style issue: add space after colon(`:`).

File: src/components/backend-ai-usersettings-general-list.ts [hunk 8]
```diff
@@ -116,14 +125,16 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+
+        .ssh-keypair {
+          margin-right:10px;
+          width: 450px;
+          min-height:100px;
+          overflow-y:scroll;
+          white-space:pre-wrap;
+          word-wrap:break-word;
+          font-size:10px;
+          scrollbar-width: none; /* firefox */
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: This is also a minor issue, but please be consistent with the space after colon(`:`). This repository prefers one space after `:`, except for inline-styles.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 9]
```diff
@@ -184,6 +205,30 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+        wl-button#ssh-keypair-details {
+          --button-bg: none;
+        }
+
+        wl-icon#ssh-keypair-icon {
+          color: var(--paper-indigo-700);
+        };
+
+        ::-webkit-scrollbar {
+          display:none; /* Chrome and Safari */
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: Here as well.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 10]
```diff
@@ -593,6 +638,93 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+  }
+
+  /**
+   *  Generate SSH Keypair By Selected method (Custom or Random)
+   */
+  _generateSSHKeypair() {
+    const generateMethodElement = this.shadowRoot.querySelector('#keypair-generation-dropdown');
+    if (generateMethodElement.selected !== null) {
+      const generateMethod = generateMethodElement.value;
+      switch (generateMethod) {
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: This will work, but wouldn't it be more reliable to use `generateMethodElement.id` instead of value? I noticed that the selector elements all have their own `id`s.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 11]
```diff
@@ -593,6 +638,93 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+        case _text("usersettings.Random") : {
+          this._refreshSSHKeypair();
+          this.shadowRoot.querySelector('#ssh-public-key').readOnly = true;
+          this.shadowRoot.querySelector('#ssh-private-key').readOnly = true;
+          break;
+        }
+        case _text("usersettings.Custom") : {
+          /**
+           * TO DO : Custom SSH Keypair registration should be applied here.
+           */
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: If this is what you asked to me in Teams, it will be good if you implement this feature.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 12]
```diff
@@ -610,6 +742,73 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+    if (this.shadowRoot.querySelector('#clear-ssh-keypair-dialog').open) {
+      this._hideSSHKeypairClearDialog();
+    }
+  }
+
+  /**
+   * Toggle readonly input in SSH keypair.
+   */
+  _toggleSSHKeypairInput() {
+    const selected: string = this.shadowRoot.querySelector('#keypair-generation-dropdown').value;
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: Here as well. For me, it does not feel good to compare elements' text description.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 13]
```diff
@@ -610,6 +742,73 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+        } else { // other browsers
+          let tmpInputElement = document.createElement("input");
+          tmpInputElement.type = "text";
+          tmpInputElement.value = copyText;
+
+          document.body.appendChild(tmpInputElement);
+          tmpInputElement.select();
+          document.execCommand("copy"); // copy operation
+          document.body.removeChild(tmpInputElement);
+        }
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts:728]: I think this part is good.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 14]
```diff
@@ -800,6 +1008,60 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+            <span slot="title"> ${_t("usersettings.CurrentSSHPublicKey")}</span>
+            <mwc-textarea class="ssh-keypair" style="width:450px;" id="current-ssh-public-key" outlined readonly></mwc-textarea>
+            <wl-button class="copy" @click="${() => this._copySSHKey("#current-ssh-public-key")}">
+              <wl-icon>content_copy</wl-icon>
+            </wl-button>
+          </div>
+          <div slot="footer">
+            <wl-button class="cancel" inverted flat @click="${this._hideSSHKeypairDialog}">${_t("button.Cancel")}</wl-button>
+            <wl-button class="ok" @click="${this._openSSHKeypairGenerationDialog}">${_t("button.Generate")}</wl-button>
+          </div>
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: The first `span` element and the following two `div` elements should be aligned with the same indent.

File: src/components/backend-ai-usersettings-general-list.ts [hunk 15]
```diff
@@ -800,6 +1008,60 @@ export default class BackendAiUsersettingsGeneralList extends BackendAIPage {
...
+                <wl-icon>content_copy</wl-icon>
+              </wl-button>
+            </div>
+            <span slot="title">${_t("usersettings.PrivateKey")}</span>
+              <div class="horizontal layout flex">
+                <mwc-textarea class="ssh-keypair" id="ssh-private-key" outlined></mwc-textarea>
+                <wl-button class="copy" @click="${() => this._copySSHKey("#ssh-private-key")}">
+                  <wl-icon>content_copy</wl-icon>
+                </wl-button>
+              </div>
```
Comment (adrysn) [inline src/components/backend-ai-usersettings-general-list.ts]: Indentation as well.

File: src/lib/backend.ai-client-node.js [hunk 16]
```diff
@@ -921,6 +921,22 @@ class Client {
             text += possible.charAt(Math.floor(Math.random() * possible.length));
         return text + "-jsSDK";
     }
+    /**
+     * fetch existing pubic key of SSH Keypair from container
+     * only ssh_public_key will be received.
+     */
+    async fetchSSHKeypair() {
+        let rqst = this.newSignedRequest('GET', `/auth/ssh-keypair`, null);
```
Comment (adrysn) [inline src/lib/backend.ai-client-node.js]: If there is no variable in the literal string, just use plain quote(`'`).
